### ISSUE TITLE
feat(async): expose coherent rollout-manager depth, lag, and phase metrics

### ISSUE THREAD

**Issue by CjhHa1** [Tencent-Hunyuan/UniRL#455]
## Context

Current `main` has replaced the request-level async scheduler from #212 with prompt-level `RolloutManager` dispatch shared by async AR and diffusion. The current manager has stronger scheduling and failure invariants, but it is difficult to answer basic runtime questions from telemetry:

- Are rollout slots saturated, queued, or starved?
- Is training blocked waiting for a rollout batch?
- How much time is spent quiescing and publishing weights?
- How much completed work is resident, carried, or discarded by the lag filter?
- Are reported policy-version metrics from one coherent point in the step?

#212 explored several useful metrics on the now-obsolete `async_runtime.py`. This issue proposes adapting the observability ideas to the current manager rather than restoring that scheduler or its `prefetch_batches` API.

### 1. Async AR/DiT bypass the phase-timing hook

[`install_phase_timing`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/utils/wandb_logger.py#L143-L162) wraps `train_step`, but [`AsyncRolloutTrainerMixin`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/trainer/async_rollout.py#L247-L280) calls `_next_rollout_batch` and `_advantage_and_train` directly. Both async trainers consequently emit only `perf/step_time_s`; they do not emit rollout-wait, reward, train, quiesce, or weight-sync phase durations.

The log call also happens inside `_advantage_and_train`, before the mixin performs a due quiesce/publication. On sync steps, `perf/step_time_s` therefore excludes mandatory post-train work and is not an end-to-end async iteration duration.

### 2. Manager residency and backpressure are not exported

[`RolloutPool.counts`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/rollout/manager/dispatch.py#L123-L129) combines queued, running, and reserved work into one value. [`RolloutManager.counts`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/rollout/manager/rollout.py#L147-L153) is used for admission but never logged.

There is no structured visibility into:

- queued versus running prompts;
- total/per-slot capacity and utilization;
- ready groups versus incomplete pending roots;
- resident backlog lag/age;
- whether admission was limited by capacity, freshness, or a hard eval/save boundary.

Without these gauges, `max_inflight`, `per_worker_inflight`, and `buffer_max_staleness` cannot be tuned from a run trace.

### 3. Drop/carry activity exists only in logs

Complete-group filtering reports a warning, and incomplete-root filtering reports an info line ([`rollout.py`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/rollout/manager/rollout.py#L92-L103), [`rollout.py`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/rollout/manager/rollout.py#L207-L217)). Suspended/carry/finish activity is likewise not counted. This makes stale-work loss and quiesce cost invisible in W&B.

### 4. One rollout record mixes pre-train and post-train versions

[`rollout_version_metrics`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/trainer/async_rollout.py#L105-L118) is computed before optimization, while [`training_version_metrics`](https://github.com/Tencent-Hunyuan/UniRL/blob/3a164b398e6368212888ed2fd72f78af90a5fc55/unirl/trainer/async_rollout.py#L121-L134) is merged after optimization.

For a batch consumed at train version 8, generated at version 4, with two optimizer updates, the same record reports:

```text
output_version=4
staleness_updates=4
train_version=10
```

Thus `train_version - output_version == 6`, while `staleness_updates == 4`. Each value is locally explainable, but the unqualified names describe different instants and make dashboards and alerts error-prone.

### A. Add telemetry-neutral manager snapshots

Expose a small typed snapshot/drain API from `RolloutManager`/`RolloutPool`; keep W&B dependencies in the trainer/logger layer.

Suggested step gauges under `rollout/async/*`:

- `queued_prompts`, `running_prompts`, `ready_groups`, `pending_roots`;
- `capacity_prompts` and aggregate slot utilization;
- oldest/max ready lag in optimizer updates or batches;
- published version.

Suggested per-step delta counters:

- submitted/launched/completed/selected prompts or groups;
- stale complete groups dropped;
- incomplete roots discarded;
- suspended, carried, and explicitly finished prompts.

Names must state whether they count prompts, logical groups, roots, or engine calls. Do not expose prompt IDs or other high-cardinality labels.

### B. Instrument explicit async phases

The async mixin should measure its actual control boundaries instead of relying on the `train_step` hook:

- `perf/rollout_collect_wait_time_s` (driver stall, not total engine runtime);
- `perf/reward_time_s`;
- `perf/train_time_s`;
- `perf/quiesce_time_s`;
- `perf/weight_sync_time_s`.

Emit timing after mandatory quiesce/publication for that step, or emit a second explicit perf record on the same rollout axis. Define whether `perf/step_time_s` includes publication, eval, and checkpoint I/O; unavailable phases should be omitted rather than logged as zero.

### C. Make version snapshots temporally coherent

Use explicit names or one snapshot boundary, for example:

- `selected_output_version`;
- `consumer_train_version` (before optimization);
- `staleness_updates = consumer_train_version - selected_output_version`;
- `train_version_after`;
- `published_version_after` and `publish_lag_after`.

Existing names can be retained during a compatibility window, but a single record must not imply an invalid arithmetic relationship.

## Acceptance criteria

- Async AR and async diffusion emit phase metrics even though they bypass `train_step`.
- A sync-due step exposes both quiesce and weight-sync duration.
- Local manager gauges distinguish queued, running, ready, and pending work without extra Ray calls.
- Drop/carry counters are visible as structured metrics, not only log messages.
- Counter reset semantics and gauge snapshot semantics are documented and tested.
- Reported version fields identify pre/post-train state, and the documented staleness equation holds.
- Focused fake-launcher tests cover normal completion, blocking collect, lag-filter drops, quiesce carry, and counter reset.
- Telemetry collection does not change admission/order/filter behavior and does not swallow manager or nested rollout failures.

## Non-goals

- Restoring `unirl/rollout/async_runtime.py` or `prefetch_batches` from #212.
- Changing current admission, FIFO, lag-filter, checkpoint, or weight-publication policy.
- Adding per-prompt labels, dashboards, or a new experiment-logger abstraction.
- Replacing detailed profiler traces or the benchmark result schema proposed in #414.
- Folding the separate agentic policy-version correctness discussion in #324 into this issue.

## Metric semantics clarifications

- `staleness_batches` is currently optimizer-update lag divided by configured `num_updates_per_batch`; skipped optimizer updates mean it is not an observed batch age. Remove it, or rename it as an explicit batch-equivalent estimate.
- Manager gauges should distinguish queued tasks, transient launch reservations, running tasks, completed-but-unrouted tasks, ready logical groups, and partial roots. Do not label `_reserved` as resident work.
- Map generic manager tasks to `*_prompts` only in async AR/DiT, where one submitted task has that meaning. Manager-owned names should remain workload-neutral.
- `admitted_prompts` should count new data-source admission only; carry resubmission after publication is a separate counter.
- Explicitly register the actual `rollout/async/*` W&B namespace. The currently registered top-level `sync/*` and `buffer/*` namespaces should not imply metrics that are never emitted.
- Add skipped-update and W&B-key binding tests so version/age semantics and the `rollout/step` axis cannot silently drift.

**Comment by coding-famer** [comment]
Hi~ I'd like to work on this if this is open to community contributor

**Comment by CjhHa1** [comment]
> Hi~ I'd like to work on this if this is open to community contributor

Yes for sure, this is still open
